tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/CircuitBreakerTest.java

```diff
@@ -44,244 +46,469 @@
+    }
+
+    /**
+     * Analagous to testCircuitClosedThenOpen but with a Class level annotation
+     * specified that is overriden by a Method level annotation on serviceC.
+     * 
+     * With successThreshold = 2, requestVolumeThreshold = 2, failureRatio=1,
+     * delay = 50000 the expected behaviour is,
+     * 
+     * Execution Behaviour 
+     * ========= ========= 
+     * 1         RunTimeException 
+     * 2         RunTimeException 
+     * 3         CircuitBreakerOpenException  
+     * 4         CircuitBreakerOpenException 
+     * 5         CircuitBreakerOpenException 
+     * 6         CircuitBreakerOpenException 
+     * 7         CircuitBreakerOpenException
+     */
+    @Test
+    public void testClassLevelCircuitOverride() {
+        for (int i = 0; i < 7; i++) {
+            try {
+                clientForClassLevelCBWithDelay.serviceC();
+
+                if (i < 2) {
+                    Assert.fail("serviceC should throw an Exception in testClassLevelCircuitOverride");
+                }
+            } 
+            catch (CircuitBreakerOpenException cboe) {
+                // Expected on iteration 4
+                if (i < 2) {
+                    Assert.fail("serviceC should throw a RuntimeException in testClassLevelCircuitOverride");
+                }
+            } 
+            catch (RuntimeException ex) {
+                // Expected
+            } 
+            catch (Exception ex) {
+                // Not Expected
+                Assert.fail("serviceC should throw a RuntimeException in testClassLevelCircuitOverride");
+            }
+        }
+
+        int serviceAExecutions = clientForClassLevelCBWithDelay.getCounterForInvokingService();
+
+        Assert.assertEquals(serviceAExecutions, 2, "The number of executions should be 2");
+    }
+
+    /**
+     * Analagous to testCircuitReClose but with a Class level annotation
+     * specified that is overriden by a Method level annotation on serviceD.
+     * 
+     * With successThreshold = 2, requestVolumeThreshold = 4, failureRatio=0.75,
+     * delay = 1 the expected behaviour is,
+     * 
+     * Execution Behaviour 
+     * ========= ========= 
+     * 1         RunTimeException 
+     * 2         RunTimeException 
+     * 3         RunTimeException  
+     * 4         RunTimeException
+     * Pause for longer than CircuitBreaker delay, so that it transitions to half-open
+     * 5         SUCCEED 
+     * 6         SUCCEED (CircuitBreaker will be re-closed as successThreshold is 2)
+     * 7         SUCCEED
+     */
+    @Test
+    public void testClassLevelCircuitOverrideNoDelay() {
+        for (int i = 0; i < 7; i++) {
+            try {
+                // Pause to allow the circuit breaker to half-open on iteration
+                // 4
+                // This is conservative when the circuit breaker has a minimal
+                // delay.
+                if (i == 4) {
+                    try {
+                        Thread.sleep(500);
```

Comment:
Instead of sleeping, I would recommend using [awaitility](https://github.com/awaitility/awaitility) to check that things are done.

Reply:
I wrote some simple wait loops that I use...

They look a little awkward in usage, but they work & they're tested unlike most adhoc wait loops:

```package core.helper.waitloop;

import java.util.function.Predicate;

import core.helper.lambdathrows.ConsumerThrows;
import core.helper.lambdathrows.FunctionThrows;

public class WaitLoopWhile<T> implements ConsumerThrows<T> {

	private FunctionThrows<T,T> whileBody;
	private Predicate<T> whileTest;
	private int count = 10;
	private int counter = 0;

	public WaitLoopWhile(Predicate<T> whileTest, FunctionThrows<T,T> whileBody) {
		this.whileTest = whileTest;
		this.whileBody = whileBody;
	}

	public WaitLoopWhile(int count, Predicate<T> whileTest, FunctionThrows<T,T> whileBody) {
		this(whileTest, whileBody);
		this.count = count;
	}

	@Override
	public void accept(T t) throws Throwable {
		while (whileTest.test(t) && ++counter <= count) {
			t = whileBody.apply(t);
		}
		if (whileTest.test(t) && counter > count) {
			throw new WaitLoopExceededException("Quit loop after " + count + " iterations.");
		}
	}
}```
...and...
```package core.helper.waitloop;

import java.util.function.Predicate;

import core.helper.lambdathrows.RunnableThrows;
import core.helper.lambdathrows.SupplierThrows;

public class WaitLoopDoWhile<T> implements RunnableThrows {

	private SupplierThrows<T> doBody;
	private Predicate<T> whileTest;
	private int count = 10;
	private int counter = 0;

	public WaitLoopDoWhile(SupplierThrows<T> doBody, Predicate<T> whileTest) {
		this.doBody = doBody;
		this.whileTest = whileTest;
	}

	public WaitLoopDoWhile(int count, SupplierThrows<T> doBody, Predicate<T> whileTest) {
		this(doBody, whileTest);
		this.count = count;
	}

	@Override
	public void run() throws Throwable {
		T t;
		do {
			t = doBody.get();
		} while (whileTest.test(t) && ++counter <= count);
		if (whileTest.test(t) && counter > count) {
			throw new WaitLoopExceededException("Quit loop after " + count + " iterations.");
		}
	}
}```